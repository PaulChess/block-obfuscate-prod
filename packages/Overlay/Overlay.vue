<template>
  <transition name="hxmui-fade" v-on:after-leave="transitionAfterLeave">
    <div
      v-show="show"
      :class="[getBemClass(), className]"
      :style="getOverlayStyle"
      @click="onClickOverlay">
      <slot></slot>
    </div>
  </transition>
</template>

<script lang="ts">
import {
  Component,
  Prop,
  Emit,
  Vue,
  Watch,
} from 'vue-property-decorator';
import {
  createNamespace,
  preventDefault,
} from '../utils/index';
import { getScroller } from '../utils/scroll';
import TouchMixin from '../mixins/touch';

const [bem] = createNamespace('overlay');

const BODY_LOCK_CLASS = 'hxm-overflow-hidden';

let totalLockCount = 0;

@Component({
  name: 'Overlay',
  mixins: [TouchMixin],
})
export default class Overlay extends Vue {
  @Prop({ default: false }) private show!: boolean;

  @Prop() private zIndex!: number | string;

  @Prop() private duration!: number | string;

  @Prop() private className!: string;

  @Prop({ default: () => {} }) private customStyle!: any;

  @Prop({ default: true }) private fixedBody!: boolean;

  @Emit('click')
  private onClickOverlay() {}

  @Watch('shouldLockBody')
  shouldLockWatcher(lock: boolean) {
    lock ? this.addBodyLock() : this.removeBodyLock();
  }

  get shouldLockBody() {
    return this.show && this.fixedBody;
  }

  get getOverlayStyle() {
    return {
      zIndex: this.zIndex !== undefined ? +this.zIndex : undefined,
      animationDuration: this.duration !== undefined ? `${this.duration}ms` : undefined,
      ...this.customStyle,
    };
  }

  private addBodyLock() {
    document.addEventListener('touchstart', this.touchStart);
    document.addEventListener('touchmove',
      this.onTouchMove,
      { passive: false });
    if (totalLockCount === 0) {
      document.body.classList.add(BODY_LOCK_CLASS);
    }
    totalLockCount++;
  }

  private removeBodyLock() {
    if (totalLockCount) {
      document.removeEventListener('touchstart', this.touchStart);
      document.removeEventListener('touchmove', this.onTouchMove);
      totalLockCount--;
      if (totalLockCount === 0) {
        document.body.classList.remove(BODY_LOCK_CLASS);
      }
    }
  }

  // https://blog.csdn.net/riddle1981/article/details/102659958
  private onTouchMove(e: TouchEvent) {
    this.touchMove(e);
    const el = getScroller(e.target as HTMLElement,
      this.$refs.popupRef as HTMLElement) as HTMLElement;
    const { scrollTop, offsetHeight, scrollHeight } = el;
    const direction = this.pos.deltaY > 0 ? '10' : '01';
    let status = '11';
    if (scrollTop === 0) {
      status = offsetHeight >= scrollHeight ? '00' : '01';
    } else if (scrollTop + offsetHeight >= scrollHeight) {
      status = '10';
    }
    if (status !== '11'
       && this.isVertical
       && !(parseInt(status, 2) & parseInt(direction, 2))) {
      if (e.cancelable) {
        preventDefault(e, true);
      }
    }
  }

  private reset() {
    totalLockCount = 0;
    document.body.classList.remove(BODY_LOCK_CLASS);
  }

  // v1.11.4 暴露overlay transition 结束的回调
  private transitionAfterLeave() {
    this.$emit('overlay-after-leave');
  }

  private getBemClass(params: any) {
    return bem(params);
  }

  beforeDestroy() {
    this.reset();
  }
}
</script>
