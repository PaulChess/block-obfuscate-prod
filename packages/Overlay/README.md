# Overlay 遮罩层 <Badge text="1.3.6+" />

## 简介

<card>

### 介绍
创建一个遮罩层，用于强调特定的页面元素，并阻止用户进行其他操作。

</card>

## 代码演示

<card>

### 基础用法

```javascript
<hxm-button @click="show = true">显示遮罩层</hxm-button>
<hxm-overlay :show="show" @click="show = false"></hxm-overlay>

export default {
  data() {
    return {
      show: false,
    };
  },
}
```
</card>

<card>

### 嵌入内容

通过默认插槽可以在遮罩层上嵌入自定义内容。

```javascript
<hxm-button @click="showEmbedded = true">嵌入内容</hxm-button>

<hxm-overlay :show="showEmbedded" @click="showEmbedded = false">
  <div class="wrapper">
    <div class="block" />
  </div>
</hxm-overlay>

export default {
  data() {
    return {
      showEmbedded: false,
    }
  }
}
```
```css
.wrapper {
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
}
.block {
  width: 2rem;
  height: 2rem;
  border-radius: 4px;
  background-color: #fff;
}
```
</card>

## API

<card>

### Props

| 参数 | 说明 | 类型 | 默认值 |
|:---|:---|:---:|:---:|
| show | 是否显示遮罩层 | boolean | `false` |
| z-index | z-index层级 | number/string | `1` |
| duration | 动画时长(单位为ms) | number/string | `300` |
| class-name | 自定义类名 | string | - |
| custom-style | 自定义样式 | object | - |
| fixedBody | 是否锁定背景滚动，锁定时蒙层中的内容也无法滚动 | boolean | `true` |

</card>

<card>

### Events

| 事件名 | 说明 | 回调参数 |
|:---:|:---:|:---:|
| click | 点击遮罩时触发 | _event: MouseEvent_ |

</card>

<card>

### Slots

| 名称 | 说明 |
|:---:|:---:|:---:|
| default | 默认插槽，用于在遮罩层上方嵌入内容 |

</card>

<demo/>
