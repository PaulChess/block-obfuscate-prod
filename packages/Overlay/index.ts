import Overlay from './Overlay.vue';
import install from '../utils/install';

export default install(Overlay);
