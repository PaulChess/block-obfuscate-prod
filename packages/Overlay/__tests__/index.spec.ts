/*
 * @Author: liushuai
 * @Date: 2021-07-15 17:34:40
 * @LastEditTime: 2021-07-19 20:57:13
 */
import { mount } from '@vue/test-utils';
import Overlay from '../Overlay.vue';

const m = (props: object = {}) => mount(Overlay, { propsData: { ...props } });
const matchSnap = (fn: (props: object) => any, props: object) => expect(fn(props).html()).toMatchSnapshot();

describe('Overlay', () => {
  /** props */
  it('props show', () => {
    matchSnap(m, { show: true });
    matchSnap(m, { show: false });
  });

  it('props slot', () => {
    matchSnap(m, {
      scopedSlots: {
        default: '<p>{{props.index}},{{props.text}}</p>',
      },
    });
  });
});
