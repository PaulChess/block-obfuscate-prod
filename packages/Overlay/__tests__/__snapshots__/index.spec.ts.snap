// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay props show 1`] = `<div class="hxm-overlay"></div>`;

exports[`Overlay props show 2`] = `<div class="hxm-overlay" style="display: none;"></div>`;

exports[`Overlay props slot 1`] = `<div class="hxm-overlay" style="display: none;" scopedslots="[object Object]"></div>`;
